providers/structured/provider.go

```diff
@@ -5,47 +5,65 @@ package structured
 
 import (
 	"context"
+	"sync"
 
 	"github.com/myhelix/contextlogger/log"
 	"github.com/myhelix/contextlogger/providers"
 	"github.com/myhelix/contextlogger/providers/chaining"
 )
 
-type RawLogCallArgs struct {
+type LogCallArgs struct {
 	ContextFields log.Fields
 	Report        bool
 	Args          []interface{}
+	Level         providers.LogLevel
 }
 
 type RecordCallArgs struct {
 	ContextFields log.Fields
 	Metrics       log.Metrics
-}
-
-type RecordEventCallArgs struct {
-	ContextFields log.Fields
 	EventName     string
-	Metrics       log.Metrics
 }
 
 type StructuredOutputLogProvider struct {
 	providers.LogProvider
 
-	rawLogCalls      map[providers.RawLogCallType][]RawLogCallArgs
-	recordCalls      []RecordCallArgs
-	recordEventCalls []RecordEventCallArgs
+	logCalls    []*LogCallArgs
+	logMutex    sync.RWMutex
+	recordCalls []*RecordCallArgs
+	recordMutex sync.RWMutex
+}
+
+// Return list of log calls, filtered to only selected levels (if any present)
+func (p StructuredOutputLogProvider) LogCalls(levels ...providers.LogLevel) (result []*LogCallArgs) {
```

Comment:
@chriswhelix I think this should return `[]LogCallArgs` instead of `[]*LogCallArgs`. We don't want to give an outside caller the ability to modify the shared underlying `LogCallArgs` object. Or alternatively, make all of `LogCallArgs'` field private, then add getters for them.

Reply:
Returning LogCallArgs instead of *LogCallArgs here doesn't accomplish that objective, though -- because the contents of the ContextFields and Args are still passed by reference even if we copy the containing struct. We'd have to go to quite a bit of extra effort to do a deep copy to really achieve that level of isolation.

Since that didn't seem worthwhile here, I figured we might as well avoid a bunch of extra copying. Also, at least this way the caller clearly knows that they're getting a shared data structure.

Reply:
I agree with this.

Reply:
I see, makes sense